### ISSUE TITLE
[unstable option] blank_lines_lower_bound

### ISSUE THREAD

**Issue by scampi** [rust-lang/rustfmt#3382]
Tracking issue for blank_lines_lower_bound

**Comment by Boscop** [comment]
It would be more useful if statements inside functions are not covered by this, only declarations!

Or is there another way to specify that I want to have at least 1 empty line between declarations (also between function-local declarations and whatever comes before/after)? (But not between statements in a function body.)

**Comment by arzg** [comment]
What are the issues blocking the stabilisation of this option?

**Comment by calebcartwright** [comment]
@arzg - The process and requirements for stabilizing a configuration option are described [here](https://github.com/rust-lang/rustfmt/blob/master/Processes.md#stabilising-an-option). Specifically, there are several issues that exist with `blank_lines_lower_bound` in released versions of rustfmt (1.x versions), for example https://github.com/rust-lang/rustfmt/issues/2954.

Some fixes/improvements to `blank_lines_lower_bound` have been made within the rustfmt source, but they have not yet been released. As such there are at least two stabilization requirements not yet for this configuration option:

> * The design and implementation of the option are sound and clean. 
> * The option is well tested, both in unit tests and, optimally, in real usage.

**Comment by Systemcluster** [comment]
With `blank_lines_lower_bound` set to `0`, rustfmt (`1.4.27-nightly (580d826e 2020-11-16)`) will still insert blank lines between `fn`s, for example

```rust
impl A {
    fn a() { /* ... */ }
    fn b() { /* ... */ }
}
```

will get formatted to

```rust
impl A {
    fn a() { /* ... */ }

    fn b() { /* ... */ }
}
```

Is this correct? I.e., should this case be covered by this configuration option?

**Comment by bluss** [comment]
This option doesn't seem to have a useful behaviour at the moment, is it something that will be tweaked?

My expectation would be that *items* (functions, struct definitions, similar) are separated by the lines (the documentation says *items*).

What actually happens is that attributes are spaced with one line before their functions, statements by one line etc, doc comments by one line before their functions etc.

```diff
@@ -1460,6 +1490,7 @@ impl<A> Clone for OwnedArcRepr<A> {
 /// [`RawArrayView`](type.RawArrayView.html) /
 /// [`RawArrayViewMut`](type.RawArrayViewMut.html) for the array type!*
 #[derive(Copy, Clone)]
+
 // This is just a marker type, to carry the mutability and element type.
 pub struct RawViewRepr<A> {
     ptr: PhantomData<A>,
@@ -1467,7 +1498,9 @@ pub struct RawViewRepr<A> {
 
 impl<A> RawViewRepr<A> {
     #[inline(always)]
+
     fn new() -> Self {
+
         RawViewRepr { ptr: PhantomData }
     }
 }
```

**Comment by calebcartwright** [comment]
> This option doesn't seem to have a useful behaviour at the moment, is it something that will be tweaked?

@bluss - your snippets are related to a bug (#2954) that's been fixed in source but not included yet in a release, it will (at a minimum) be corrected to only ensure spacing between items

**Comment by Keavon** [comment]
Hello, would it be possible to please get a update on the status of this being stabilized? Thank you very much.

**Comment by dcow** [comment]
When 

> > This option doesn't seem to have a useful behaviour at the moment, is it something that will be tweaked?
> 
> @bluss - your snippets are related to a bug (#2954) that's been fixed in source but not included yet in a release, it will (at a minimum) be corrected to only ensure spacing between items

I am still experiencing this bug.

**Comment by ytmimi** [comment]
@dcow I'd suggest that you open up a new issue to fully describe the problem that you're experiencing

**Comment by dxps** [comment]
Any updates on this, please? :pray: 
To improve the readability I love to have a blank line between a "complex"/lengthy signature of a function and its body. 
So, I am very interested on using this.
Thanks!

**Comment by dxps** [comment]
> I would say it is pretty common that teams want blank newlines between function definitions

Yes, that would be preferrable, as some functions signature becomes readable when a new blank line within the function body exists.

**Comment by azizur** [comment]
I am new to Rust world. The fact that it's over 5 years since issue was raised. I am thinking this project has been abandoned?

**Comment by calebcartwright** [comment]
> I am new to Rust world.

Welcome :wave: 

>  The fact that it's over 5 years since issue was raised. I am thinking this project has been abandoned?

No, and I'd suggest that using the amount of time that has passed since any given issue was opened as a measure for whether a project is active vs. abandoned is not going to lead to accurate results.

I don't think there's any singular way to adequately gauge a project's health, but I think some better indicators would be things like whether maintainers are responding, whether there are some issues being opened/triaged/closed, (in some cases) whether there's still code changes being made, bugs being fixed/features being added, etc.

I'd posit that if you go look at most large-scale projects on GitHub you'll find they have open issues that are much, much older than 5 years!

**Comment by joshtriplett** [comment]
If I understand correctly, https://github.com/rust-lang/rustfmt/pull/4295 is fixed, but the fix can't be applied to the default style edition because it'd be a formatting change, right?

@calebcartwright If we applied this fix to the next style edition, would that then unblock the ability to stabilize this option?

**Comment by calebcartwright** [comment]
I don't have the full context on this top of mind, but as I recall there's a lot of open questions around this option and whether the existing behavior as it currently exists is even what we want (not to mention that I'm not confident the option is currently bug free nor has met the [criteria for stabilization](https://github.com/rust-lang/rustfmt/discussions/5365))

**Comment by estebank** [comment]
@joshtriplett from rereading the threads (and extrapolating from the behavior I'm seeing on my codebase when enabling the feature), https://github.com/rust-lang/rustfmt/pull/4295#issuecomment-1131665173 fixing https://github.com/rust-lang/rustfmt/issues/2954 was merged against rustfmt-2.0, but never backported against the released rustfmt. Hence why https://github.com/rust-lang/rustfmt/pull/5438 was created.
